Sources/Apollo/HTTPNetworkTransport.swift

```diff
@@ -104,110 +112,160 @@ public class HTTPNetworkTransport: NetworkTransport {
-                                 error: receivedError,
-                                 for: request,
-                                 response: response,
-                                 completionHandler: completionHandler)
+        self.handleErrorOrRetry(operation: operation,
+                                error: receivedError,
+                                for: request,
+                                response: response,
+                                completionHandler: completionHandler)
         return
       }
-
+      
       guard let httpResponse = response as? HTTPURLResponse else {
         fatalError("Response should be an HTTPURLResponse")
       }
-
+      
       guard httpResponse.isSuccessful else {
         let unsuccessfulError = GraphQLHTTPResponseError(body: data,
                                                          response: httpResponse,
                                                          kind: .errorResponse)
-        self?.handleErrorOrRetry(operation: operation,
-                                 error: unsuccessfulError,
-                                 for: request,
-                                 response: response,
-                                 completionHandler: completionHandler)
+        self.handleErrorOrRetry(operation: operation,
+                                error: unsuccessfulError,
+                                for: request,
+                                response: response,
+                                completionHandler: completionHandler)
         return
       }
-
+      
       guard let data = data else {
         let error = GraphQLHTTPResponseError(body: nil,
                                              response: httpResponse,
                                              kind: .invalidResponse)
-        self?.handleErrorOrRetry(operation: operation,
-                                 error: error,
-                                 for: request,
-                                 response: response,
-                                 completionHandler: completionHandler)
+        self.handleErrorOrRetry(operation: operation,
+                                error: error,
+                                for: request,
+                                response: response,
+                                completionHandler: completionHandler)
         return
       }
-
+      
       do {
-        guard let body = try self?.serializationFormat.deserialize(data: data) as? JSONObject else {
+        guard let body = try self.serializationFormat.deserialize(data: data) as? JSONObject else {
           throw GraphQLHTTPResponseError(body: data, response: httpResponse, kind: .invalidResponse)
         }
-        let response = GraphQLResponse(operation: operation, body: body)
-        completionHandler(response, nil)
+        if self.enableAutoPersistedQueries,
+          let error = body["errors"] as? [JSONObject],
```

Comment:
this is getting bit messy here... How about creating `GraphQLResponse` (else path) and adding method to it to get errors. That way you don't need to know here much about `JSONObject` and `GraphQLResponse` already knows about it.

```swift
let response = GraphQLResponse(operation: operation, body: body)

if response.errors.contain(...) {

} else {
   completionHandler(response, nil)
}

Reply:
good idea, will update accordingly!